tictactoe/game1.py

```diff
@@ -54,46 +62,128 @@ def is_draw(self):
+
+
+@dc.dataclass
+class Update(Action):
+    tile: int
+
+    def execute(self, game: Game):
+        b = game.board
+        if b[self.tile] == Mark._:
+            no_of_marked_tiles = game.marked_tiles()
             n_even_marked_tiles = no_of_marked_tiles % 2 == 0
-            b[tile] = repr(Mark.x) if n_even_marked_tiles else repr(Mark.o)
-            self.filled_tiles.append(tile)
-            print(self)
+            b[self.tile] = Mark.x if n_even_marked_tiles else Mark.o
+            game.filled_tiles.append(self.tile)
+            print(game)
         else:
             print("tile already picked")
 
-    def undo(self):
-        if self.marked_tiles() != 0:
-            self.board[self.filled_tiles[-1]] = Mark._
-            self.filled_tiles.pop()
-            print(self)
+
+class Undo(Action):
+    def execute(self, game: Game):
+        if game.marked_tiles() != 0:
+            game.board[game.filled_tiles[-1]] = Mark._
+            game.filled_tiles.pop()
+            print(game)
         else:
             print("cant undo")
 
 
-class main:
-    filled_tiles = []
-    game = Game(filled_tiles)
-    print(game)
+class StorageDriver(abc.ABC):
+    @abc.abstractmethod
+    def store_move(self, update: Update): ...
+    @abc.abstractmethod
+    def read_moves(self) -> list[Update]: ...
+    @abc.abstractmethod
+    def undo_move(self): ...
+    def delete_moves(self): ...
+
+
+@dc.dataclass
+class StoreDB(StorageDriver):
+    def __post_init__(self):
+        self.con = sqlite3.connect("tictactoe.db")
+        self.cur = self.con.cursor()
+        self.cur.execute(
+            "CREATE TABLE IF NOT EXISTS moves(id INTEGER PRIMARY KEY, tile INTEGER)"
+        )
+        self.con.commit()
+
+    def store_move(self, update: Update):
+        self.cur.execute("INSERT INTO moves(tile) VALUES (?)", (update.tile,))
+        self.con.commit()
+
+    def read_moves(self) -> list[Update]:
+        res = self.cur.execute("SELECT tile FROM moves")
+        moves = res.fetchall()
+        print("fetchingg", moves)
+        print("listttt", [Update(move[0]) for move in moves])
+        return [Update(move[0]) for move in moves]
+
+    def undo_move(self):
+        self.cur.execute("DELETE FROM moves WHERE id = (SELECT MAX(id) FROM moves)")
+        self.con.commit()
+
+    def delete_moves(self):
+        self.cur.execute("DELETE FROM moves")
+        self.con.commit()
+
+
+@dc.dataclass
+class DBUpdate(Action):
+    sd: StoreDB
+    update: Update
+
+    def execute(self, game: Game):
+        self.sd.store_move(self.update)
+        self.update.execute(game)
+
+
+@dc.dataclass
+class DBUndo(Action):
+    sd: StoreDB
+    undo: Undo
+
+    def execute(self, game: Game):
+        self.sd.undo_move()
+        self.undo.execute(game)
+
+
+def main():
+    game = Game()
+
+    sd = StoreDB()
+    for update in sd.read_moves():
+        game.apply(update)
+
     while True:
         input1 = int(input("pick a tile "))
```

Comment:
instead of using the `input` builtin function directly make an interface and its implementation class that will read from a variety of sources i.e. `stdin`, `file`, `database`, `network` etc.